components/function-controller/pkg/apis/serverless/v1alpha2/function_defaults.go

```diff
@@ -94,24 +94,55 @@ func (spec *FunctionSpec) defaultReplicas(config *DefaultingConfig, fn *Function
 }
 
 func (spec *FunctionSpec) defaultFunctionResources(config *DefaultingConfig, fn *Function) {
-	resources := spec.ResourceConfiguration.Function.Resources
+	var resources *corev1.ResourceRequirements
+	if spec.ResourceConfiguration != nil && spec.ResourceConfiguration.Function != nil && spec.ResourceConfiguration.Function.Resources != nil {
+		resources = spec.ResourceConfiguration.Function.Resources
+	}
 	defaultingConfig := config.Function.Resources
 	resourcesPreset := mergeResourcesPreset(fn, FunctionResourcesPresetLabel, defaultingConfig.Presets, defaultingConfig.DefaultPreset, defaultingConfig.RuntimePresets)
+	calculatedResources := defaultResources(resources, resourcesPreset.RequestMemory, resourcesPreset.RequestCPU, resourcesPreset.LimitMemory, resourcesPreset.LimitCPU)
+
+	if spec.ResourceConfiguration == nil {
+		spec.ResourceConfiguration = &ResourceConfiguration{}
+	}
+
+	if spec.ResourceConfiguration.Function == nil {
+		spec.ResourceConfiguration.Function = &ResourceRequirements{}
+	}
 
-	spec.ResourceConfiguration.Function.Resources = defaultResources(&resources, resourcesPreset.RequestMemory, resourcesPreset.RequestCPU, resourcesPreset.LimitMemory, resourcesPreset.LimitCPU)
+	spec.ResourceConfiguration.Function.Resources = calculatedResources
 }
 
 func (spec *FunctionSpec) defaultBuildResources(config *DefaultingConfig, fn *Function) {
-	buildResourceCfg := spec.ResourceConfiguration.Build
+
 	// if build resources are not set by the user we don't default them.
 	// However, if only a part is set or the preset label is set, we should correctly set missing defaults.
 	if shouldSkipBuildResourcesDefault(fn) {
 		return
 	}
 
+	var buildResourceCfg *ResourceRequirements
+	if spec.ResourceConfiguration != nil && spec.ResourceConfiguration.Build != nil {
+		buildResourceCfg = spec.ResourceConfiguration.Build
+	} else {
+		buildResourceCfg = &ResourceRequirements{}
+		buildResourceCfg.Resources = &corev1.ResourceRequirements{}
+	}
+
 	defaultingConfig := config.BuildJob.Resources
 	resourcesPreset := mergeResourcesPreset(fn, BuildResourcesPresetLabel, defaultingConfig.Presets, defaultingConfig.DefaultPreset, nil)
-	spec.ResourceConfiguration.Build.Resources = defaultResources(&buildResourceCfg.Resources, resourcesPreset.RequestMemory, resourcesPreset.RequestCPU, resourcesPreset.LimitMemory, resourcesPreset.LimitCPU)
+
+	calculatedResources := defaultResources(buildResourceCfg.Resources, resourcesPreset.RequestMemory, resourcesPreset.RequestCPU, resourcesPreset.LimitMemory, resourcesPreset.LimitCPU)
+
+	if spec.ResourceConfiguration == nil {
+		spec.ResourceConfiguration = &ResourceConfiguration{}
+	}
+
+	if spec.ResourceConfiguration.Build == nil {
+		spec.ResourceConfiguration.Build = &ResourceRequirements{}
+	}
+
+	spec.ResourceConfiguration.Build.Resources = calculatedResources
```

Comment:
We can put it to seperate functon like `setBuildResources`

Reply:
done